modules/core/src/main/scala/doobie/util/fragment.scala

```diff
@@ -24,156 +23,209 @@ object fragment {
-
-    /** Construct an [[Update0]] from this fragment. */
-    @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
-    def update(implicit h: LogHandler = LogHandler.nop): Update0 =
-      updateWithLogHandler(h)
-
-    /** Construct an [[Update0]] from this fragment with the given `LogHandler`. */
-    def updateWithLogHandler(h: LogHandler): Update0 =
-      Update[elems.type](sql, pos, h).toUpdate0(elems)
-
-    override def toString =
-      s"""Fragment("$sql")"""
-
-    /** Used only for testing; this pulls out the arguments as an untyped list. */
-    private def args: List[Any] =
-      elems.toList.map {
-        case Elem.Arg(a, _) => a
-        case Elem.Opt(a, _) => a
-      }
-
-    /** Used only for testing; this uses universal equality on the captured arguments. */
-    @SuppressWarnings(Array("org.wartremover.warts.Equals"))
-    private[util] def unsafeEquals(fb: Fragment): Boolean =
-      sql == fb.sql && args == fb.args
-
   }
-  object Fragment {
 
     /**
      * Construct a statement fragment with the given SQL string, which must contain sufficient `?`
      * placeholders to accommodate the given list of interpolated elements. This is normally
      * accomplished via the string interpolator rather than direct construction.
      */
     @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
-    def apply(sql: String, elems: List[Elem], pos: Option[Pos] = None): Fragment =
+    def apply[R](sql: String,
+                 elems: List[Elem[R]],
+                 pos: Option[Pos] = None): Fragment[R] =
       new Fragment(sql, Chain.fromSeq(elems), pos)
 
     /**
      * Construct a statement fragment with no interpolated values and no trailing space; the
      * passed SQL string must not contain `?` placeholders.
      */
     @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
-    def const0(sql: String, pos: Option[Pos] = None): Fragment =
+    def const0(sql: String, pos: Option[Pos] = None): Fragment[Any] =
       new Fragment(sql, Chain.empty, pos)
 
     /**
      * Construct a statement fragment with no interpolated values and a trailing space; the
      * passed SQL string must not contain `?` placeholders.
      */
      @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
-     def const(sql: String, pos: Option[Pos] = None): Fragment =
+    def const(sql: String, pos: Option[Pos] = None): Fragment[Any] =
       const0(sql + " ", pos)
 
     /** The empty fragment. Adding this to another fragment has no effect. */
-    val empty: Fragment =
+    val empty: Fragment[Any] =
       const0("")
 
     /** Statement fragments form a monoid. */
-    implicit val FragmentMonoid: Monoid[Fragment] =
-      new Monoid[Fragment] {
-        val empty = Fragment.empty
-        def combine(a: Fragment, b: Fragment) = a ++ b
+    implicit def FragmentMonoid[R]: Monoid[Fragment[R]] =
+      new Monoid[Fragment[R]] {
```

Comment:
Since `R` is in a phantom position, might be worth caching+casting this instance to avoid constructing more than one.